### PR TITLE
Add PID controller for j2s7s300 moveit trajectory execution

### PR DESCRIPTION
**Pull request by sjhansen3**
## Here is a trajectory controller that has been built in python. It wraps a trajectory controller our lab uses in an action server. 

I am able to get a significant improvement in accuracy of the robot's trajectories. Tolerances are consistently under 5mm in task space :)

I would however not recommend accepting this pull request. It still has issues:

First off:
I don't understand the surrounding driver code to be honest, I ended up working around much of it enrirely and send velocity commands myself. I do not pass FollowJointTrajectoryFeedback to /move_group - I just report FollowJointTrajectoryResult and leave it at that. I don't do nearly as much input checking as the code I replaced does. I just wanted to get something built quickly to test. This has only been tested on the j2s7s300 - I don't support any other robotic arms.

**More importantly:**
I am get occasional time outs - **I would like advice or recommendations on this.**

I am running into an issue which I believe is related to self collisions. I believe moveit develops a plan using the URDF while the jaco has a low level collision checker which prevents motion in a certain range - so sometimes I cannot reach my goal because the goal is infeasible. 

Basically I am concerned that the planner is not as conservative as it should be.

If this is an issue related to how conservative the planner is I believe I can fix it by adding padding to the links. Of course this is tricky - too much padding and we wont be able to find paths to configurations that are in fact feasible. It also looks like padding is not a fully implemented feature for moveit.

https://github.com/ros-planning/moveit/issues/683

## Here are 5 cases where one joint failed to get within the goal tolerance.
Failure 1:
[INFO] [WallTime: 1518404273.280986] current joint angles: [[ 43.62971115]
[150.4420929 ]
[488.00985718]
[239.08143616]
[214.39830017]
[ 65.04006958]
[388.21832275]]
[INFO] [WallTime: 1518404273.282013] Not reached goal current distance: [[6.40830388e-06]
[3.02244493e-04]
[2.12908228e-04]
[2.51760364e-04]
[4.39682447e-04]
[1.52755051e-02]
[1.48752425e-04]]

Failure 2
[INFO] [WallTime: 1518405606.281880] current joint angles: [[ 49.56307602]
[147.1570282 ]
[317.427948 ]
[121.01952362]
[ 18.2250042 ]
[ 65.02274323]
[377.88095093]]
[INFO] [WallTime: 1518405606.382141] Not reached goal current distance: [[1.91015640e-05]
[1.53024171e-04]
[2.01943631e-04]
[4.76082113e-06]
[4.18909020e-05]
[9.38528001e-02]
[2.74310224e-05]]

Failure 3
[INFO] [WallTime: 1518405778.080664] current joint angles: [[ -5.06790018]
[ 240.74482727]
[ 67.55264282]
[ 120.79373932]
[-153.08294678]
[ 294.828125 ]
[ 202.52116394]]
[INFO] [WallTime: 1518405778.281218] Not reached goal current distance: [[3.96728317e-05]
[5.42856740e-05]
[1.07109230e-04]
[1.46091427e-04]
[1.76913295e-04]
[6.57344447e-02]
[5.59733724e-05]]

Failure 4
[INFO] [WallTime: 1518408206.856982] current joint angles: [[ 354.58374023]
[ 241.5188446 ]
[-115.34962463]
[ 239.10058594]
[ 29.25539398]
[ 294.14404297]
[ 205.46577454]]
[INFO] [WallTime: 1518408206.858735] Not reached goal current distance: [[2.53245914e-04]
[2.56339710e-04]
[9.42463080e-06]
[1.09646227e-05]
[4.70486963e-04]
[5.08751173e-02]
[3.71398264e-05]]

Here are 5 cases where Joint 6 was not able to get within the goal tolerance (off by 0.05-0.09 radians)

## I have a few questions I would like clarification on:
1. Does collision checking happen after velocity commands have been sent to the robot?
2. How does this collision checking work? Is it more or less conservative than the collision checking used by moveit?
3. Is there anywhere in the Kinova code that padding is added?

Thanks
-Steve

### CODE REVIEW

**Comment by martine1406** [comment]
Dear Steve

Thank you for your work. I have some elements of answers for you

The urdf limits for joint 6 in kinova_description/urdf/j2s7s300.xacro are mistaken. Angle limits for this joint on the real robot are 65 degrees and 295 degrees. I will change that in the urdf. I will also check the other joint limits to insure theu match the real robot's ones. 

to answer your questions: 
1- Does collision checking happen after velocity commands have been sent to the robot?. In angular mode, the only collision check that will be done is on joint limits. 
2- How does this collision checking work? Is it more or less conservative than the collision checking used by moveit? no, I would even say the ones on moveit would be more conservative as the real robot is controlled in angular mode in this case. If the real robot would be controlled in Cartesian mode (i.e. receive Cartesian command from MoveIt, it would be different. 
3- Is there anywhere in the Kinova code that padding is added? I do not think so. Do you mean in the Kinova arm firmware? 

I will test your pull request today and let you know
Again, I big thanks for your work
Martine

**Comment by martine1406** [comment]
Hi Steve

I had a look at the files you added and modified in your pull request. In file kinova_driver/src/kinova_joint_trajectory_controller.cpp, are you sure you have not removed some of the valuable changes you had made in commit c77623c when you did commit fe9a865? Because after commit fe9a865, the for loop **for ( int i=0; i<cur_joint_position.size(); ++i )** and the threshold EPSILON_GOAL seem pretty useless. Let me know

File: kinova_driver/src/kinova_joint_trajectory_controller.cpp [hunk 1]
```diff
@@ -219,25 +236,22 @@ void JointTrajectoryController::pub_joint_vel(const ros::TimerEvent&)
 //                          std::endl <<" joint_velocity_msg.joint5: " << joint_velocity_msg.joint5 * M_PI/180 <<
 //                          std::endl <<" joint_velocity_msg.joint6: " << joint_velocity_msg.joint6 * M_PI/180 );
 
-        pub_joint_velocity_.publish(joint_velocity_msg);
+        //pub_joint_velocity_.publish(joint_velocity_msg);
```
Comment (sjhansen3) [inline kinova_driver/src/kinova_joint_trajectory_controller.cpp]: The publish is commented out. This controller doesn't do anything. 

I know this is rather confusing, but I was to concerned to rip the whole controller out, so I just prevented it from publishing joint velocity commands.

File: kinova_driver/src/scripts/pid_controller.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        else:
+            t = time.time() - self.path_start_T
+            self.target_pos = self.interpolate_trajectory(t)
+
+            if not self.reached_goal:
+                #TODO add a timeout if the goal is taking too long
+                dist_from_goal = PIDController.shortest_angular_distance(curr_pos, self.goal)
+                rospy.loginfo_throttle(5, "Not reached goal current distance: {}".format(np.abs(dist_from_goal)))
+
+                is_at_goal = np.all(np.abs(dist_from_goal) < GOAL_EPSILON)
```
Comment (sjhansen3) [inline kinova_driver/src/scripts/pid_controller.py]: This is where I check to see if I have reached the goal.

**Comment by sjhansen3** [comment]
Hi Martine,

I added a few comments to clarify. kinova_driver/src/kinova_joint_trajectory_controller.cpp no longer publishes commands. I commented out the line which publishes velocity commands. Ultimately I will remove the control code entirely. The goal check in kinova_driver/src/kinova_joint_trajectory_controller.cpp was from an older attempt to integrate the controller.

**Comment by sjhansen3** [comment]
I just added commit 15d5a4a8d11ce0ebbf43419672dc63b0e68d8f36 which removes the unneeded code. Is there any update on the URDF? I think it would explain a lot if the URDF joint limits are set incorrectly.

**Comment by martine1406** [comment]
Hi Steve

are you getting this error?

[ERROR] [1518472697.433512062]: Controller is taking too long to execute trajectory (the expected upper bound for the trajectory execution was 5.755021 seconds). Stopping trajectory.

I get some successful trajectory executions, but most of them end up with this error

**Comment by sjhansen3** [comment]
yes that is precisely the error that I am having. I believe the cause is related to incorrect joint limits. Essentially I think that the robot is trying to reach a location that the robot cannot reach. 

For debug you can also turn off moveit's trajectory time-out using these instructions:
https://answers.ros.org/question/196586/how-do-i-disable-execution_duration_monitoring/

You can see in my PR I changed the allowed_execution_duration_scaling to 4

**Comment by sjhansen3** [comment]
Here are several configurations that I planned to and their result as to whether they timed out or not.

Success = No time-out, goal achieved
Failure = The trajectory timed out

Success
---
joint1: 25.7683601379
joint2: 277.355072021
joint3: 734.387329102
joint4: 239.277633667
joint5: 250.975402832
joint6: 255.20652771
joint7: 156.375274658
---

Success
---
joint1: 116.931503296
joint2: 182.809570312
joint3: 887.703674316
joint4: 121.291404724
joint5: 279.548858643
joint6: 261.8828125
joint7: 222.215484619
---

Success
---
joint1: 14.7830657959
joint2: 278.093170166
joint3: 708.976989746
joint4: 238.951751709
joint5: 273.341888428
joint6: 269.882507324
joint7: 154.986495972
---

Success
---
joint1: 100.974128723
joint2: 132.610549927
joint3: 537.88482666
joint4: 239.050292969
joint5: 286.300964355
joint6: 279.005249023
joint7: 156.62059021
---

Success
---
joint1: 193.138427734
joint2: 82.2787094116
joint3: 709.130615234
joint4: 121.202339172
joint5: 453.952545166
joint6: 268.434356689
joint7: 154.712921143
---

Success
---
joint1: 320.41229248
joint2: 218.129501343
joint3: 578.469848633
joint4: 121.218551636
joint5: 400.079376221
joint6: 239.450271606
joint7: 168.591186523
---

Fail - none seem to be at limit??
---
joint1: 218.734359741
joint2: 74.1314468384
joint3: 852.408325195
joint4: 125.175094604
joint5: -37.160369873
joint6: 250.420730591
joint7: 309.142456055
---

Fail - Joint 2 certainly at limit, very nearly hitting itself
---
joint1: 217.436676025
joint2: 46.640625
joint3: 581.982055664
joint4: 237.545501709
joint5: 240.89630127
joint6: 257.64175415
joint7: 453.854949951
---

**Comment by martine1406** [comment]
Hi Steve

I think that even allowed_execution_duration_scaling to 4.0 is not enough in some circumstances, in particular when the start state and the goal_state are relatively close. With corrected angle limits in the URDF and with allowed_execution_duration_scaling set to 10.0 (or with <param name="trajectory_execution/execution_duration_monitoring" value="false" /> as recommended in the link you sent me https://answers.ros.org/question/196586/how-do-i-disable-execution_duration_monitoring/), I do not get the error anymore. Maybe we have to increase the PID values in pid.py so the robot moves faster? In all cases, if someone is willing to sacrifice execution speed  for a better accuracy, your work seems to do the trick. I'll continue testing tomorrow and maybe create a special development branch with your fixes in them.

**Comment by martine1406** [comment]
* correction, PID gains have to bu updated in pid_controller.py, not pid.py. I'll test this

File: kinova_driver/src/scripts/pid_controller.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+
+        # stores maximum COMMANDED joint torques
+        self.max_cmd = MAX_CMD_VEL * np.eye(7)
+        # stores current COMMANDED joint torques
+        self.cmd = np.eye(7)
+        # stores current joint MEASURED joint torques
+        self.joint_torques = np.zeros((7, 1))
+
+        # P, I, D gains
+        p_gain = 60.0
```
Comment (sjhansen3) [inline kinova_driver/src/scripts/pid_controller.py]: Can you elaborate about your concern Martine? I see no issue with setting PID gains here.

**Comment by sjhansen3** [comment]
Can you provide the new .xacro files you are using with correct joint limits?

I think you could be right regarding the trade off between accuracy and speed. We might also need to tune the gains better or increase the GOAL_EPSILON to allow for positions that are further from the goal.

I want to make sure I am using the correct joint limits first though so please provide the correct URDF.

Thanks!

**Comment by martine1406** [comment]
Hi Steve

I pushed the new .xacro files with the right angle limits for all robots yesterday on kinova-ros master branch. On my tests yesterday, I stopped getting this error when I increase P and D by factor 1.5 and I used an I of 0.2 (and with the new xacro files too). I did not tune the PID gains extensively, so I probably did not reached the best compromise. Let me know if you have time to test and reach a better solution. 

Thanks again for your hard work. Very useful for the community
Martine

**Comment by martine1406** [comment]
I tested more and I do not get the <Controller is taking too long to execute trajectory> error anymore with higher PID gains. However, my robot mode in MoveIt is moving more slowly than my real robot. I did not get this behaviour with the current kinova-ros master branch. I wonder why....

**Comment by martine1406** [comment]
ok it is most probably because of the feedback on the trajectory execution that is not yet fully implemented

**Comment by martine1406** [comment]
made a new branch dev/improved_precision_moveit_plan that implement the changes made in this pull request along with other changes to make this work useful for other robots in addition to the 7 dof one.

**Comment by Kukanani** [comment]
I've been trying to get this branch to work with a 7DOF Jaco, but the robot moves extremely slowly while executing trajectories from MoveIt. @sjhansen3 are you using MoveIt or some other trajectory planner?

**Comment by sjhansen3** [comment]
@Kukanani I am using MoveIt. How are you creating plans?

There are a few things that I can think of that could cause the robot to move slow:
1. The trajectory controller is not very tuned. You can edit the gains here https://github.com/Kinovarobotics/kinova-ros/blob/869a267c0d5895a014101fb63c9a117a0c3c59df/kinova_driver/src/scripts/pid_controller.py#L88

**Comment by Kukanani** [comment]
I should have been more specific, sorry. The robot starts moving promptly, but moves extremely slowly (takes 10-30 seconds to complete a simple linear move). The console output suggests that the plan itself is short, but most of the motion happens after the trajectory time has elapsed because of the low speed.

As a result, the robot is interpolating from close to the start directly to the end point of the trajectory, which does not result in good path following behavior.

Our lab also has a somewhat functional trajectory controller written in C++. It's hard to make a PR for because we're so diverged at this point, but I'm working on it. We use velocity control directly from the trajectory, and once I put a position PID over the top of it the follow behavior should be pretty good. Alternatively, I'm thinking of implementing a controller stack based on existing ROS packages (i.e. ros_control).

**Comment by martine1406** [comment]
Hi Stephen

I cannot merge this pull request with the master branch because it seems to change some files that were configured for the 7 dof robot and make them compatible with a 6 dof robot instead. Besides, we prefer merging pull requests from the community in our beta branch, not the master one (kinova-ros-beta branch). 
This work seems to be in line with your previous work on using the kinova-ros package on kinetic as well (correct me if I am wrong). 
In all cases, if you can do another pull request on the kinova-ros-beta, it would be much appreciated. Besides, would it be possible to modify your last commit d79d27ab352554372afbe455b0bb96696e49ec74 so that it does not overwrites the files that were written for jaco 7 dof? Or else, I do not think I would merge this commit 

thank you
Martine

**Comment by sjhansen3** [comment]
This was not intended to be merged. I didnt know the PR was open. We can close this PR. I was getting my controller to work with the Mico and yes it was pretty crappy in its implementation. It basically broke much of what I did previously.

**Comment by martine1406** [comment]
ok perfect. Thanks for the heads up. Good luck in your work!

**Pull request closed without merging**